src/hotspot/share/services/memoryService.cpp

```diff
@@ -202,6 +202,21 @@ bool MemoryService::set_verbose(bool verbose) {
   return verbose;
 }
 
+bool MemoryService::get_verbose() {
+  for (LogTagSet* ts = LogTagSet::first(); ts != nullptr; ts = ts->next()) {
+    // set_verbose only sets gc and not gc*, so check for an exact match
+    const bool is_gc_exact_match = ts->contains(LogTag::_gc) && ts->ntags() == 1;
+    if (is_gc_exact_match) {
+      LogLevelType l = ts->level_for(LogConfiguration::StdoutLog);
+      if (l == LogLevel::Info || l == LogLevel::Debug || l == LogLevel::Trace) {
+        return true;
+      }
+    }
+  }
+
+  return false;
+}
```

Comment:
Update: Looks like the CSR says that's exactly what we want.

Reply:
Wait. This is not what the implementation does. If you specify -Xlog:gc and -Xlog:gc+init (using a real tag instead of foo), `get_verbose` will return true. It only cares about the tag set that is exactly 'gc' and ignores the 'gc+init' tag set.

I'm adding a test line to show that.

Reply:
So in this test case:
    `-Xlog:gc,gc+init`
you have it so that `true` is returned.

With this line of code:
 ```
   const bool is_gc_exact_match = ts->contains(LogTag::_gc) && ts->ntags() == 1;
```
I would thing that this part is `false`: `ts->ntags() == 1`
for the `-Xlog:gc,gc+init` test case so we'll be returning `false`.

What am I missing?

Reply:
The important part is that `ts` contains one single tag set and that `gc,gc+init` is a string that specifies the two tag sets 'gc' and 'gc+init'. Also, note that `-Xlog:gc,gc+init` and `-Xlog:gc -Xlog:gc+init` is effectively the same.

It is also important to understand that the for loop iterates over all tag sets that have been used in the HotSpot code. It is not iterating over the tag sets specified by -Xlog lines.

Reply:
OK, I get it now. The loop cycles thru the tag set and finds one that is
a singleton and an exact match, i.e., `-Xlog:gc`. With `-Xlog:gc+init`,
the `ts->ntags() == 1` would be `false` because that one is not an
exact match.

I misunderstood the algorithm a bit. Sorry for the noise.